### PR TITLE
HOCS-2196: Multiple Contribution Somu Endpoints

### PR DESCRIPTION
**Pull request by LiamMacP**
This pull requests adds the following endpoints so that they can be used with cases to add somu items: 

- Delete:                       /item/{somuUuid}
- Creation/ Update:     /case/{caseUuid}/item/{somuTypeUuid}
- Get 1:                         /case/{caseUuid}/item/{somuTypeUuid} 
- Get Cases:                /case/{caseUuid}/item

Tests have also be added for the Resource, Service, DTO's and Models. This also includes our interactions with the audit service.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/SomuItemResource.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public class SomuItemResource {
+
+    private final SomuItemService somuItemService;
+
+    @Autowired
+    public SomuItemResource(SomuItemService somuItemService) {
+        this.somuItemService = somuItemService;
+    }
+    
+    @GetMapping(value = "/case/{caseUuid}/item")
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/SomuItemResource.java:25]: Would be better to tell it that it produces JSON. Same for line 31.

Comment (Paulcr-ho) [reply src/main/java/uk/gov/digital/ho/hocs/casework/api/SomuItemResource.java:25]: Agreed - I will add it.

**Review by MarkWrightBjss — COMMENTED** [review]
Generally, I prefer unit tests to compare against static values where possible. (Minor point, and no need to retro-fit in this PR.)

**Comment by LiamMacP** [comment]
> Generally, I prefer unit tests to compare against static values where possible. (Minor point, and no need to retro-fit in this PR.)

Yeah I totally get that, will do that moving forward.

**Comment by Paulcr-ho** [comment]
After discussing the responsibilities of the backend and frontend node server, we decided that the backend service would be responsible for dealing with the database update or creation (the latest change to this PR), while the node server would be responsible for the contents of the json data with arrays of items.

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/dto/GetSomuItemResponseTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public void getSomuItemResponseTest() {
+        SomuItem somuItem = new SomuItem(SOMU_ITEM_UUID, SOMU_ITEM_CASE_UUID, SOMU_ITEM_TYPE_UUID, "");
+        
+        GetSomuItemResponse getSomuItemResponse = GetSomuItemResponse.from(somuItem);
+
+        assertThat(getSomuItemResponse.getUuid()).isEqualTo(somuItem.getUuid());
+        assertThat(getSomuItemResponse.getCaseUuid()).isEqualTo(somuItem.getCaseUuid());
+        assertThat(getSomuItemResponse.getSomuUuid()).isEqualTo(somuItem.getSomuUuid());
+        assertThat(getSomuItemResponse.getData()).isEqualTo(somuItem.getData());
+        assertThat(getSomuItemResponse.isDeleted()).isEqualTo(somuItem.isDeleted());
```
[the remark this answers is not in the document]
Comment (LiamMacP) [reply src/test/java/uk/gov/digital/ho/hocs/casework/api/dto/GetSomuItemResponseTest.java]: Good shout, this is now rectified 👍🏼

File: src/test/java/uk/gov/digital/ho/hocs/casework/api/dto/GetSomuItemResponseTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    public void getSomuItemResponseTest_NullDataIsDeleted() {
+        SomuItem somuItem = new SomuItem(SOMU_ITEM_UUID, SOMU_ITEM_CASE_UUID, SOMU_ITEM_TYPE_UUID, null);
+
+        GetSomuItemResponse getSomuItemResponse = GetSomuItemResponse.from(somuItem);
+
+        assertThat(getSomuItemResponse.getUuid()).isEqualTo(somuItem.getUuid());
+        assertThat(getSomuItemResponse.getCaseUuid()).isEqualTo(somuItem.getCaseUuid());
+        assertThat(getSomuItemResponse.getSomuUuid()).isEqualTo(SOMU_ITEM_TYPE_UUID);
+        assertThat(getSomuItemResponse.getData()).isEqualTo(somuItem.getData());
+        assertThat(getSomuItemResponse.isDeleted()).isEqualTo(somuItem.isDeleted());
```
[the remark this answers is not in the document]
Comment (LiamMacP) [reply src/test/java/uk/gov/digital/ho/hocs/casework/api/dto/GetSomuItemResponseTest.java]: Good shout, this is now rectified 👍🏼

File: src/main/java/uk/gov/digital/ho/hocs/casework/client/auditclient/AuditClient.java [hunk 4]
```diff
@@ -233,6 +233,65 @@ public void deleteCaseNoteAudit(CaseNote caseNote) {
...
+        RequestDataDto requestDataDto = RequestDataDto.from(requestData);
+        LocalDateTime localDateTime = LocalDateTime.now();
+        executorService.execute(() -> {
+            String data = "{}";
+            try {
+                data = objectMapper.writeValueAsString(new AuditPayload.SomuItem(somuItem.getUuid(), somuItem.getCaseUuid(), somuItem.getSomuUuid(), somuItem.getData()));
+            } catch (JsonProcessingException e) {
+                logFailedToParseDataPayload(e);
+            }
+            sendAuditMessage(localDateTime, somuItem.getCaseUuid(), data, EventType.SOMU_ITEM_CREATED, null,
```
Comment (LiamMacP) [inline src/main/java/uk/gov/digital/ho/hocs/casework/client/auditclient/AuditClient.java:260]: We need to change the calls to use the sendAuditMessage that includes the data in the audit data, that will allow for search to add it to indexing on Elastic Search. 

This needs changing here and in the update and delete commands.

Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/casework/client/auditclient/AuditClient.java]: This is now resolved.

**Pull request merged**